src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3354,38 +3354,70 @@ private static Symbol AsMemberOfType(NamedTypeSymbol containingType, Symbol symb
+                int? memberIndex = symbol.Kind == SymbolKind.Property ? symbol.MemberIndexOpt : null;
+                if (!memberIndex.HasValue)
                 {
-                    if (symbolDefContainer.IsTupleType)
+                    return symbol;
+                }
+                return AnonymousTypeManager.GetAnonymousTypeProperty(containingType, memberIndex.GetValueOrDefault());
+            }
+            if (!symbolDefContainer.IsGenericType)
+            {
+                Debug.Assert(symbol.ContainingType.IsDefinition);
+                return symbol;
+            }
+            if (symbolDefContainer.IsInterface)
+            {
+                if (tryAsMemberOfSingleType(containingType, out Symbol result))
+                {
+                    return result;
+                }
+                foreach (var @interface in containingType.AllInterfacesNoUseSiteDiagnostics)
+                {
+                    if (tryAsMemberOfSingleType(@interface, out result))
                     {
-                        return AsMemberOfTupleType((TupleTypeSymbol)containingType, symbol);
+                        return result;
                     }
-                    if (symbolDefContainer.IsAnonymousType)
+                }
+            }
+            else
+            {
+                while (true)
+                {
+                    if (tryAsMemberOfSingleType(containingType, out Symbol result))
                     {
-                        int? memberIndex = symbol.Kind == SymbolKind.Property ? symbol.MemberIndexOpt : null;
-                        if (!memberIndex.HasValue)
-                        {
-                            break;
-                        }
-                        return AnonymousTypeManager.GetAnonymousTypeProperty(containingType, memberIndex.GetValueOrDefault());
+                        return result;
                     }
-                    var result = symbolDef.SymbolAsMember(containingType);
-                    if (result is MethodSymbol resultMethod && resultMethod.IsGenericMethod)
+                    containingType = containingType.BaseTypeNoUseSiteDiagnostics;
+                    if ((object)containingType == null)
                     {
-                        return resultMethod.Construct(((MethodSymbol)symbol).TypeArguments);
+                        break;
                     }
-                    return result;
                 }
-                containingType = containingType.BaseTypeNoUseSiteDiagnostics;
-                if ((object)containingType == null)
+            }
+            Debug.Assert(false); // If this assert fails, add an appropriate test.
+            return symbol;
+
+            bool tryAsMemberOfSingleType(NamedTypeSymbol singleType, out Symbol result)
+            {
+                if (!singleType.OriginalDefinition.Equals(symbolDefContainer, TypeCompareKind.AllIgnoreOptions))
```

Comment:
>TypeCompareKind.AllIgnoreOptions [](start = 78, length = 32)

Do we want to match tuple names and perhaps dynamic?

Reply:
Perhaps although that doesn't seem necessary.

---
In reply to: [261332320](https://github.com/dotnet/roslyn/pull/33764#discussion_r261332320) [](ancestors = 261332320)